libtenzir/src/table_slice.cpp

```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
+  }
+}
+
+auto realize(unflatten_record&& record) -> std::shared_ptr<arrow::StructArray>;
+
+auto realize(unflatten_entry&& entry) -> std::shared_ptr<arrow::Array> {
+  TENZIR_WARN("pre realize");
+  return entry.match<std::shared_ptr<arrow::Array>>(
+    [](std::shared_ptr<arrow::Array>& array) {
+      TENZIR_ASSERT(array);
+      TENZIR_WARN("realizing {}", array->type()->ToString());
+      return std::move(array);
+    },
+    [](unflatten_record& record) {
+      TENZIR_WARN("realizing record");
+      return realize(std::move(record));
+    });
+}
+
+auto realize(unflatten_record&& record) -> std::shared_ptr<arrow::StructArray> {
+  TENZIR_WARN("realizing record now");
+  auto names = std::vector<std::string>{};
+  auto arrays = std::vector<std::shared_ptr<arrow::Array>>{};
+  for (auto& [name, entry] : record.fields) {
+    TENZIR_WARN("seeing {}", name);
+    names.emplace_back(name);
+    arrays.push_back(realize(std::move(entry)));
+  }
+  return make_struct_array(record.length, record.null_bitmap, std::move(names),
+                           arrays);
+}
+
+auto unflatten(const arrow::StructArray& array, std::string_view sep)
+  -> std::shared_ptr<arrow::StructArray> {
+  // { foo.bar.baz: 1, foo.bar.qux: 2, foo: {bar: 42} }
+  auto root = unflatten_entry{unflatten_record{array.length(), nullptr}};
+  unflatten_recurse(root, array, sep);
+  auto record = std::get_if<unflatten_record>(&root);
+  TENZIR_ASSERT(record);
+  return realize(std::move(*record));
+}
+
+auto unflatten(std::shared_ptr<arrow::Array> array, std::string_view sep)
+  -> std::shared_ptr<arrow::Array>;
+
+auto unflatten(const arrow::ListArray& array, std::string_view sep)
+  -> std::shared_ptr<arrow::ListArray> {
+  // TODO: array.offset?
+  auto values = unflatten(array.values(), sep);
+  return check(arrow::ListArray::FromArrays(
+    *array.offsets(), *values, arrow::default_memory_pool(),
+    array.null_bitmap(), array.data()->null_count));
+}
+
+auto unflatten(std::shared_ptr<arrow::Array> array, std::string_view sep)
+  -> std::shared_ptr<arrow::Array> {
+  if (auto record = caf::get_if<arrow::StructArray>(&*array)) {
+    return unflatten(*record, sep);
+  }
+  if (auto list = caf::get_if<arrow::ListArray>(&*array)) {
+    return unflatten(*list, sep);
+  }
+  return array;
+}
+
+auto unflatten2(const table_slice& slice, std::string_view sep) -> table_slice {
+  if (slice.rows() == 0) {
+    // TODO: I don't like this.
+    return slice;
+  }
+  auto array = check(to_record_batch(slice)->ToStructArray());
+  auto result = unflatten(array, sep);
+  auto cast = std::dynamic_pointer_cast<arrow::StructArray>(std::move(result));
+  TENZIR_ASSERT(cast);
+  auto schema = type{slice.schema().name(), type::from_arrow(*cast->type())};
+  // TODO: Is this what we want?
+  auto batch = arrow::RecordBatch::Make(schema.to_arrow_schema(),
+                                        cast->length(), cast->fields());
+  // TODO: Offset, import_time, etc.
```

Comment:
Especially the import time is rather important for the compaction use case, given that that strongly relies on it.

I think the offset we could actually remove, it should be dead code by now.